test/e2e/healthcheck_test.go

```diff
@@ -0,0 +1,215 @@
+				err := k8sClient.Patch(ctx, dnsRecord, patchFrom)
+				g.Expect(err).To(BeNil())
+			}, TestTimeoutMedium, time.Second).Should(Succeed())
+
+			By("Confirming the DNS Record status")
+			Eventually(func(g Gomega) {
+				err := k8sClient.Get(ctx, client.ObjectKeyFromObject(dnsRecord), dnsRecord)
+				g.Expect(err).ToNot(HaveOccurred())
+
+				if healthChecksSupported {
+					g.Expect(dnsRecord.Status.HealthCheck).ToNot(BeNil())
+					g.Expect(&dnsRecord.Status.HealthCheck.Probes).ToNot(BeNil())
+					g.Expect(len(dnsRecord.Status.HealthCheck.Probes)).ToNot(BeZero())
+					for _, condition := range dnsRecord.Status.HealthCheck.Conditions {
+						if condition.Type == "healthProbesSynced" {
+							g.Expect(condition.Status).To(Equal(metav1.ConditionTrue))
+							g.Expect(condition.Reason).To(Equal("AllProbesSynced"))
+						}
+					}
+				} else {
+					g.Expect(dnsRecord.Status.HealthCheck).ToNot(BeNil())
+					g.Expect(dnsRecord.Status.HealthCheck.Probes).To(BeNil())
+				}
+
+				for _, probe := range dnsRecord.Status.HealthCheck.Probes {
+					g.Expect(probe.Host).To(Equal(testHostname))
+					g.Expect(probe.IPAddress).To(Equal("172.32.200.1"))
+					g.Expect(probe.ID).ToNot(Equal(""))
+
+					for _, probeCondition := range probe.Conditions {
+						g.Expect(probeCondition.Type).To(Equal("ProbeSynced"))
+						g.Expect(probeCondition.Status).To(Equal(metav1.ConditionTrue))
+						g.Expect(probeCondition.Message).To(ContainSubstring(fmt.Sprintf("id: %v, address: %v, host: %v", probe.ID, probe.IPAddress, probe.Host)))
+					}
+				}
+			}, TestTimeoutMedium, time.Second).Should(Succeed())
+
+			By("confirming the health checks exist in the provider")
+			Eventually(func(g Gomega) {
+				if !healthChecksSupported {
+					g.Expect(len(dnsRecord.Status.HealthCheck.Probes)).To(BeZero())
+				}
+				for _, healthCheck := range dnsRecord.Status.HealthCheck.Probes {
+					exists, err := provider.HealthCheckReconciler().HealthCheckExists(ctx, &healthCheck)
+					g.Expect(err).To(BeNil())
+					g.Expect(exists).To(BeTrue())
+				}
+			}, TestTimeoutMedium, time.Second).Should(Succeed())
+
+			By("Deleting the DNS Record")
+			oldHealthCheckStatus = dnsRecord.Status.HealthCheck.DeepCopy()
+			err = helpers.ResourceFromFile("./fixtures/healthcheck_test/geo-dnsrecord-healthchecks.yaml", dnsRecord, helpers.GetTestEnv)
+			Expect(err).ToNot(HaveOccurred())
+			Eventually(func(g Gomega) {
+				err := k8sClient.Delete(ctx, dnsRecord)
+				g.Expect(client.IgnoreNotFound(err)).To(BeNil())
+
+				err = k8sClient.Get(ctx, client.ObjectKeyFromObject(dnsRecord), dnsRecord)
+				g.Expect(errors.IsNotFound(err)).Should(BeTrue())
+			}, TestTimeoutMedium, time.Second).Should(Succeed())
+
+			By("confirming the health checks were removed in the provider")
+			Eventually(func(g Gomega) {
+				for _, healthCheck := range oldHealthCheckStatus.Probes {
+					exists, err := provider.HealthCheckReconciler().HealthCheckExists(ctx, &healthCheck)
```

Comment:
What was wrong with that it was passing when the health didn't exist?

Reply:
it jumped straight over to the status check, and I hadn't added the following:
```
g.Expect(len(dnsRecord.Status.HealthCheck.Probes)).ToNot(BeZero())
```

So it looped over all 0 probes and thought it was OK.

Reply:
the problem wasn't in the `healthcheckExists though, it was [here](https://github.com/Kuadrant/dns-operator/pull/113/files#diff-8e289e1d78e0d1d3a21094f8373fad5460cae17ca8f70d2ce7de1721e2d7cc00R76) and [here](https://github.com/Kuadrant/dns-operator/pull/113/files#diff-8e289e1d78e0d1d3a21094f8373fad5460cae17ca8f70d2ce7de1721e2d7cc00R155)

Reply:
ok, so it never executed this before?

Reply:
yeah ok, since you squashed it was hard to tell what you changed to make it work. Get it now.